Automattic/newspack-plugin
# feat(registration-block): newsletter subscription

### All Submissions:






### Changes proposed in this Pull Request:

Implements newsletter subscription support for the Registration Block as provided by Newspack Newsletters ESP-agnostic API.

The block supports 2 style variations but the stacked version is prioritized on smaller viewports.

<img width="709" alt="image" src="https://user-images.githubusercontent.com/820752/179255485-12ce5287-7bf7-4672-98e5-343ad406a6c4.png">

## Stacked layout

### Multiple newsletters with narrow width

This is also the default layout for the mobile viewport.

<img width="447" alt="image" src="https://user-images.githubusercontent.com/820752/179255736-c4042787-32fa-4193-b55a-58a029de7216.png">

### Multiple newsletters with wider width

<img width="888" alt="image" src="https://user-images.githubusercontent.com/820752/179256067-b96e4d66-a88f-48f0-9215-d745eaaf5988.png">

### Single newsletter

The checkbox label is editable to represent a more generic newsletter subscription messaging.

<img width="691" alt="image" src="https://user-images.githubusercontent.com/820752/179256261-a5bcdc20-f6f0-4224-90de-f6df305146a4.png">

## Column layout

This layout is only displayed on the desktop viewport and with newsletter subscriptions enabled.

### Multiple newsletters

<img width="1078" alt="image" src="https://user-images.githubusercontent.com/820752/179256628-038665a7-fe89-4957-892d-40495b51a104.png">

### Single newsletter

<img width="1065" alt="image" src="https://user-images.githubusercontent.com/820752/179256731-03b05f8c-0b29-4409-9d48-f5a400a4af62.png">



Closes #1738

### How to test the changes in this Pull Request:

1. Make sure you have AMP Plus and Reader Activation enabled (`define( 'NEWSPACK_EXPERIMENTAL_READER_ACTIVATION', true );`)
2. Also confirm you have newsletter subscription lists configured on Newspack Newsletters (Engagement wizard)
3. Place a Registration Block on a new page and enable Newsletter subscription
4. With one newsletter selected, confirm you can update the newsletter checkbox label through the block's rich text editor
5. Select multiple newsletters and confirm the UI is updated
6. Save the block using the default "Stacked" layout and test it on the frontend
7. In an incognito window, confirm the registration works and the email is subscribed to the selected lists on your ESP
8. Change your layout to "Column" and confirm the layout is as expected
9. Register again and confirm it also works as expected

Note: I intend to tackle the success message layout next, not in this PR.

#### Editable text

Confirm you can edit the privacy text, "Newsletters" title (displayed on Column layout with multiple newsletters), and Button label through the in-block rich text editor.

### Other information:
